test/integration/id-map.test.ts

```diff
@@ -0,0 +1,403 @@
+import { expect, fetchMock } from "../test-helper"
+import { ApplicationRecord, Author, Book, Bio } from "../fixtures"
+import { IDMap } from "../../src/id-map"
+
+afterEach(() => {
+  ApplicationRecord.store.data = {}
+})
+
+describe("ID Map", () => {
+  beforeEach(() => {
+    ApplicationRecord.sync = true
+    let responsePayload = (firstName: string) => {
+      return {
+        data: {
+          id: "1",
+          type: "authors",
+          attributes: { firstName }
+        }
+      }
+    }
+
+    fetchMock.get(
+      "http://example.com/api/v1/authors/1",
+      responsePayload('John')
+    )
+
+    fetchMock.delete("http://example.com/api/v1/authors/1", {
+      meta: {}
+    })
+
+    fetchMock.delete("http://example.com/api/books/1", {
+      meta: {}
+    })
+
+    fetchMock.delete("http://example.com/api/bios/1", {
+      meta: {}
+    })
+  })
+
+  afterEach(() => {
+    fetchMock.reset()
+  })
+
+  describe("when fetching from server", () => {
+    describe("when disabled", () => {
+      beforeEach(() => {
+        ApplicationRecord.sync = false
+      })
+
+      it("does not add to the idmap", async () => {
+        let { data } = await Author.find(1)
+        let stored = ApplicationRecord.store.data
+        expect(Object.keys(stored).length).to.eq(0)
+      })
+    })
+
+    it("is added to the ID map", async () => {
+      let { data } = await Author.find(1)
+      let stored = ApplicationRecord.store.data
+      expect(Object.keys(stored)[0]).to.eq('authors-1')
+      expect(stored['authors-1']).to.deep.eq(data.attributes)
+    })
+
+    // A good way to think about this test:
+    //
+    // On the one hand, we want instances to sync - imagine a page
+    // that has a grid on the left, and a detail view on the right
+    // when you click a grid record. Both should always reflect the
+    // same global state.
+    //
+    // On the other hand, imagine that detail view was instead a form -
+    // we wouldn't want the user typing in the form to update the grid,
+    // until the form was saved.
+    it("syncs multiple instances", async () => {
+      fetchMock.put("http://example.com/api/v1/authors/1", {
+        data: {
+          id: "1",
+          type: "authors",
+          attributes: {
+            firstName: "updated"
+          }
+        }
+      })
+
+      let response = await Author.find(1)
+      let author1 = response.data
+      response = await Author.find(1)
+      let author2 = response.data
+
+      // not synced prior to save
+      author1.firstName = 'updated'
+      expect(author2.firstName).to.not.eq('updated')
+
+      await author1.save()
+
+      // now synced after save
+      expect(author2.firstName).to.eq('updated')
+
+      // now back to unsynced
+      author1.firstName = 'updated again'
+      expect(author2.firstName).to.eq('updated')
```

Comment:
Don’t have to do it for every case of this, but I would recommend adding one or two smoke tests around calling `unlisten()` not causing further changes, both if the unlistened model saves itself and if another saves.